libopenage/options.h

```diff
@@ -0,0 +1,291 @@
+
+
+using opt_func_t = std::function<void()>;
+
+// TODO string description
+// This is a base class for event_class, event, action_t mapped to a function
+// This could be constructed with a context and bind itself using the required types
+class OptionAction {
+public:
+	OptionAction(const std::string &name, const opt_func_t f);
+
+	/**
+	 * pass mouse position
+	 */
+	bool do_action();
+
+	/**
+	 * name of this action
+	 */
+	const std::string name;
+
+private:
+	const opt_func_t function;
+
+};
+
+/**
+ * Allows objects to expose customisable features
+ *
+ * This is used to bind inner game classes
+ * with console interaction or gui elements
+ */
+class OptionNode {
+	template<class T>
+	friend class Var;
+public:
+	OptionNode(const std::string &panel_name);
+	virtual ~OptionNode();
+
+	/**
+	 * lists all available options in a readable format
+	 */
+	std::vector<std::string> list_options(bool recurse=false, std::string indent="");
+
+	/**
+	 * shows all available variable names
+	 */
+	std::vector<std::string> list_variables() const;
+
+	/**
+	 * shows all the available function names
+	 */
+	std::vector<std::string> list_functions() const;
+
+	OptionValue &get_variable_rw(const std::string &name);
+	const OptionValue &get_variable(const std::string &name) const;
+
+	/**
+	 * shortcut for get_variable(name).value<T>()
+	 */
+	template<class T>
+	const T &getv(const std::string &name) {
+		return this->get_variable(name).value<T>();
+	}
+
+	/**
+	 * return a child by name, or null if not available
+	 */
+	OptionNode *get_child(const std::string &name) const;
+
+	/**
+	 * Set a parent panel which
+	 * implies adding a child to the parent
+	 */
+	void set_parent(OptionNode *parent);
+
+	/**
+	 * performs the named action
+	 */
+	void do_action(const std::string &aname);
+
+	/**
+	 * name of this node
+	 */
+	const std::string name;
+
+protected:
+
+	/**
+	 * add types to the interface
+	 */
+	void add(const std::string &vname, const OptionValue &value);
+
+	void add_action(const OptionAction &action);
+
+private:
+
+	/**
+	 * add child nodes
+	 */
+	void add_panel(OptionNode *child);
+
+	/**
+	 * remove child nodes
+	 */
+	void remove_panel(OptionNode *child);
+
+	/**
+	 * parent of this node
+	 */
+	OptionNode *parent;
+
+	/**
+	 * Variables which can be used
+	 * TODO: read only variables
+	 */
+	std::unordered_map<std::string, OptionValue> varmap;
+
+
+	// list available functions for interaction
+	std::unordered_map<std::string, OptionAction> actions;
+
+	/**
+	 * children of this node
+	 */
+	std::unordered_map<std::string, OptionNode *> children;
+};
+
+
+/**
+ * A interaface variable which gets monitored by an
+ * option node allowing reflection, while also
+ * being directly accessable as a typed member
+ */
+template<class T>
+class Var : public util::Variable<T> {
+public:
+	Var(OptionNode *owner, const std::string &name, const T &init)
+		:
+		util::Variable<T>{init} {
+		owner->add(name, &this->value);
```

Comment:
warning: address of 'this->value' will always evaluate to 'true' [-Wpointer-bool-conversion]